### ISSUE TITLE
Link error on windows 10 building with visual studio 2022

### ISSUE THREAD

**Issue by EternalSaga** [yuzu-emu/yuzu#8551]
When I was building the yuzu master branch (commit 313f047f974249c0fa004056ced3f18a8c61eae4), linker report the error  
````
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol __gmtime_r referenced in function ASN1_TIME_adj_internal	yuzu-cmd	D:\ProgramAndStudy\Cpp_proj\yuzu\build\src\yuzu_cmd\crypto-49.lib(a_time_tm.obj)	1	
````

It seems like compiler is trying to link a system library on linux.

**Comment by Docteh** [comment]
What leads you to believe its a system library on linux? 

`__gmtime_r`  is the symbol its not finding, that doesn't seem overly linux to me.

**Comment by EternalSaga** [comment]
> What leads you to believe its a system library on linux?
> 
> `__gmtime_r` is the symbol its not finding, that doesn't seem overly linux to me.

I‘m not sure it's a linux library, but I found it on this site. [](https://linux.die.net/man/3/gmtime_r). I have never used time.h before but it seems is a ansi c library. It should be found on windows.   
I'm trying to use yuzu-mainline and build it again.

**Comment by EternalSaga** [comment]
Strange, the cpp reference site reveals that the gmtime_r is a C23 function. [cppreference](https://en.cppreference.com/w/c/chrono/gmtime) If it is ture, it's quite possible that msvc has not implemented it yet.

**Comment by Docteh** [comment]
Looking at your log line, I'd guess that maybe something different with crypto got pulled in somehow. 

Like with D:\ProgramAndStudy\Cpp_proj\yuzu\build\src\yuzu_cmd\crypto-49.lib(a_time_tm.obj) does that mean that `D:\ProgramAndStudy\Cpp_proj\yuzu\build\src\yuzu_cmd\crypto-49.lib` is a file in there? I did a Debug build of yuzu_cmd and build/src/yuzu_cmd only has cmake and visual studio project files, other files are in build/src/yuzu_cmd/yuzu-cmd.dir/Debug

**Comment by EternalSaga** [comment]
Ok, I got the same error with yuzu mainline.
````
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK2019	unresolved external symbol __gmtime_r referenced in function ASN1_TIME_adj_internal	yuzu-cmd	D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\src\yuzu_cmd\crypto-49.lib(a_time_tm.obj)	1	
Error	LNK2001	unresolved external symbol __gmtime_r	yuzu-cmd	D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\src\yuzu_cmd\crypto-49.lib(a_time.obj)	1	
Error	LNK1120	1 unresolved externals	yuzu-cmd	D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\bin\Release\yuzu-cmd.exe	1	
Error	LNK2019	unresolved external symbol __gmtime_r referenced in function ASN1_TIME_adj_internal	yuzu	D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\src\yuzu\crypto-49.lib(a_time_tm.obj)	1	
Error	LNK2001	unresolved external symbol __gmtime_r	yuzu	D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\src\yuzu\crypto-49.lib(a_time.obj)	1	
Error	LNK1120	1 unresolved externals	yuzu	D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\bin\Release\yuzu.exe	1	

````

**Comment by EternalSaga** [comment]
> Looking at your log line, I'd guess that maybe something different with crypto got pulled in somehow.
> 
> Like with D:\ProgramAndStudy\Cpp_proj\yuzu\build\src\yuzu_cmd\crypto-49.lib(a_time_tm.obj) does that mean that `D:\ProgramAndStudy\Cpp_proj\yuzu\build\src\yuzu_cmd\crypto-49.lib` is a file in there? I did a Debug build of yuzu_cmd and build/src/yuzu_cmd only has cmake and visual studio project files, other files are in build/src/yuzu_cmd/yuzu-cmd.dir/Debug

However, there is not a file crypto-49.lib exists in the folder D:\ProgramAndStudy\Cpp_proj\yuzu\build\src\yuzu_cmd.

**Comment by Docteh** [comment]
please upload your build\src\yuzu_cmd\yuzu-cmd.vcxproj I wish to see what it says for crypto-49.lib

As part of AdditionalDependencies
..\..\externals\libressl\crypto\Debug\crypto-49.lib

**Comment by EternalSaga** [comment]
> please upload your build\src\yuzu_cmd\yuzu-cmd.vcxproj I wish to see what it says for crypto-49.lib
> 
> As part of AdditionalDependencies ....\externals\libressl\crypto\Debug\crypto-49.lib

@Docteh Thank you very much.
````xml
<?xml version="1.0" encoding="utf-8"?>
<Project DefaultTargets="Build" ToolsVersion="17.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <PropertyGroup>
    <PreferredToolArchitecture>x64</PreferredToolArchitecture>
  </PropertyGroup>
  <ItemGroup Label="ProjectConfigurations">
    <ProjectConfiguration Include="Debug|x64">
      <Configuration>Debug</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
    <ProjectConfiguration Include="Release|x64">
      <Configuration>Release</Configuration>
      <Platform>x64</Platform>
    </ProjectConfiguration>
  </ItemGroup>
  <PropertyGroup Label="Globals">
    <ProjectGuid>{69F7D34C-BFFE-3B59-A9C9-D0C143383D01}</ProjectGuid>
    <Keyword>Win32Proj</Keyword>
    <WindowsTargetPlatformVersion>10.0.19041.0</WindowsTargetPlatformVersion>
    <Platform>x64</Platform>
    <ProjectName>RUN_TESTS</ProjectName>
    <VCProjectUpgraderObjectName>NoUpgrade</VCProjectUpgraderObjectName>
  </PropertyGroup>
  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'" Label="Configuration">
    <ConfigurationType>Utility</ConfigurationType>
    <CharacterSet>MultiByte</CharacterSet>
    <PlatformToolset>v143</PlatformToolset>
  </PropertyGroup>
  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'" Label="Configuration">
    <ConfigurationType>Utility</ConfigurationType>
    <CharacterSet>MultiByte</CharacterSet>
    <PlatformToolset>v143</PlatformToolset>
  </PropertyGroup>
  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
  <ImportGroup Label="ExtensionSettings">
    <Import Project="$(VCTargetsPath)\BuildCustomizations\masm.props" />
  </ImportGroup>
  <ImportGroup Label="PropertySheets">
    <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" Label="LocalAppDataPlatform" />
  </ImportGroup>
  <PropertyGroup Label="UserMacros" />
  <PropertyGroup>
    <_ProjectFileVersion>10.0.20506.1</_ProjectFileVersion>
    <IntDir Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">$(Platform)\$(Configuration)\$(ProjectName)\</IntDir>
    <IntDir Condition="'$(Configuration)|$(Platform)'=='Release|x64'">$(Platform)\$(Configuration)\$(ProjectName)\</IntDir>
  </PropertyGroup>
  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">
    <PostBuildEvent>
      <Message></Message>
      <Command>setlocal
"C:\Program Files\CMake\bin\ctest.exe" --force-new-ctest-process -C $(Configuration)
if %errorlevel% neq 0 goto :cmEnd
:cmEnd
endlocal &amp; call :cmErrorLevel %errorlevel% &amp; goto :cmDone
:cmErrorLevel
exit /b %1
:cmDone
if %errorlevel% neq 0 goto :VCEnd</Command>
    </PostBuildEvent>
  </ItemDefinitionGroup>
  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
    <PostBuildEvent>
      <Message></Message>
      <Command>setlocal
"C:\Program Files\CMake\bin\ctest.exe" --force-new-ctest-process -C $(Configuration)
if %errorlevel% neq 0 goto :cmEnd
:cmEnd
endlocal &amp; call :cmErrorLevel %errorlevel% &amp; goto :cmDone
:cmErrorLevel
exit /b %1
:cmDone
if %errorlevel% neq 0 goto :VCEnd</Command>
    </PostBuildEvent>
  </ItemDefinitionGroup>
  <ItemGroup>
    <CustomBuild Include="D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\CMakeFiles\9cf5e504fdc1186deb0c2c550f41f424\RUN_TESTS_force.rule">
      <Message Condition="'$(Configuration)|$(Platform)'=='Debug|x64'"> </Message>
      <Command Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">setlocal
cd .
if %errorlevel% neq 0 goto :cmEnd
:cmEnd
endlocal &amp; call :cmErrorLevel %errorlevel% &amp; goto :cmDone
:cmErrorLevel
exit /b %1
:cmDone
if %errorlevel% neq 0 goto :VCEnd</Command>
      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">%(AdditionalInputs)</AdditionalInputs>
      <Outputs Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\src\yuzu_cmd\CMakeFiles\RUN_TESTS_force</Outputs>
      <LinkObjects Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">false</LinkObjects>
      <VerifyInputsAndOutputsExist Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">false</VerifyInputsAndOutputsExist>
      <Message Condition="'$(Configuration)|$(Platform)'=='Release|x64'"> </Message>
      <Command Condition="'$(Configuration)|$(Platform)'=='Release|x64'">setlocal
cd .
if %errorlevel% neq 0 goto :cmEnd
:cmEnd
endlocal &amp; call :cmErrorLevel %errorlevel% &amp; goto :cmDone
:cmErrorLevel
exit /b %1
:cmDone
if %errorlevel% neq 0 goto :VCEnd</Command>
      <AdditionalInputs Condition="'$(Configuration)|$(Platform)'=='Release|x64'">%(AdditionalInputs)</AdditionalInputs>
      <Outputs Condition="'$(Configuration)|$(Platform)'=='Release|x64'">D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\src\yuzu_cmd\CMakeFiles\RUN_TESTS_force</Outputs>
      <LinkObjects Condition="'$(Configuration)|$(Platform)'=='Release|x64'">false</LinkObjects>
      <VerifyInputsAndOutputsExist Condition="'$(Configuration)|$(Platform)'=='Release|x64'">false</VerifyInputsAndOutputsExist>
    </CustomBuild>
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="D:\ProgramAndStudy\Cpp_proj\yuzu-mainline\build\ZERO_CHECK.vcxproj">
      <Project>{16BD96BA-33FA-36D9-B284-A2AC0CE1A230}</Project>
      <Name>ZERO_CHECK</Name>
      <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
      <CopyToOutputDirectory>Never</CopyToOutputDirectory>
    </ProjectReference>
  </ItemGroup>
  <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
  <ImportGroup Label="ExtensionTargets">
    <Import Project="$(VCTargetsPath)\BuildCustomizations\masm.targets" />
  </ImportGroup>
</Project>
````

**Comment by Docteh** [comment]
Hmm that part looks fine.

build\externals\libressl there is LibreSSL.sln, maybe open that and set it to Release (might compile faster) and ask it to Rebuild the ALL_BUILD solution on the right hand side? 

From the initial error its not clear if the problem is with MAKING crypto-49.lib or USING it

**Comment by EternalSaga** [comment]
> Hmm that part looks fine.
> 
> build\externals\libressl there is LibreSSL.sln, maybe open that and set it to Release (might compile faster) and ask it to Rebuild the ALL_BUILD solution on the right hand side?
> 
> From the initial error its not clear if the problem is with MAKING crypto-49.lib or USING it

Unfortunately, this method still not help for the problem.  
I tried to relplace gmtime_r function to gmtime and build it again.

**Comment by Docteh** [comment]
> Unfortunately, this method still not help for the problem.

Was it actually able to build libressl without a problem? Goal there was to see if it was with MAKING `crypto-49.lib` or USING it


Not sure if it's relevant, but due to #8349 well, the issues that caused that reversion, I'm using a slightly older version of Visual Studio 2022 17.1.5 It's possible things got weirder?

**Comment by EternalSaga** [comment]
> > Unfortunately, this method still not help for the problem.
> 
> Was it actually able to build libressl without a problem? Goal there was to see if it was with MAKING `crypto-49.lib` or USING it
> 
> Not sure if it's relevant, but due to #8349 well, the issues that caused that reversion, I'm using a slightly older version of Visual Studio 2022 17.1.5 It's possible things got weirder?

Yes, crypto-49.lib can be compiled successfully and it is located in the folder yuzu-mainline\build\externals\libressl\crypto\Release.  
My vs2022 version is the latest 17.2.5 but I don't know how to roll back the version of visual studio.

**Comment by EternalSaga** [comment]
> Visual Studio Installer should let you install Visual Studio Community 2019 along side. 2022.
> 
> Got the extra 1.7GB on C:?
> 
> I think your option with VS2022 is 17.1.5 or not use VS2022.
> 
> ~Just remember to change the generator option over in cmake~
> 
> Turns out you [should not just change the generator](https://stackoverflow.com/questions/32569800/changing-current-cmake-generator) that stack overflow post suggests clearing out the build dir and starting over.
> 
> I have three build directories because of that.
> 
> Latest version of MSVC2019 should work fine, that is what is being used to generate the mainline builds.

@Docteh 
Thank you for your guide.   
I use msvc2019 and tried it again. However, the error still exists.  That's really weird.
![error](https://user-images.githubusercontent.com/15174504/178004319-d16469ce-d649-4b97-86d2-bb0c465e75d0.PNG)

**Comment by Docteh** [comment]
TL;DR I have no idea why that linker error shows up.

From what I can tell __gmtime_r comes from libressl
https://github.com/citra-emu/ext-libressl-portable/blob/8929f818fd748fd31a34fec7c04558399e13014a/crypto/compat/timegm.c#L191-L201

Right click a term and right click, for the "peek definition" option.

With a_time_tm.c in the libressl I started with peek definition for gmtime_r it's in D:\github\yuzu\externals\libressl\include\compat\time.h
![image](https://user-images.githubusercontent.com/190571/178087931-af615e40-6a7f-4f6c-b910-f427d26ca8ce.png)


Then when I peek `__gmtime_r` from time.h I get D:\github\yuzu\externals\libressl\crypto\compat\timegm.c
![image](https://user-images.githubusercontent.com/190571/178088316-1c8e6af8-2759-468d-8a41-2cf3ee45173f.png)

Looks like a proper definition


My first thought was actually to peek at the  crypto-49.lib file to see what gets exported, and __gmtime_r seems to be exported
build\externals\libressl\crypto\Release> dumpbin crypto-49.lib -linkermember > linkermember.txt
700kb of text, no big deal.

```
Microsoft (R) COFF/PE Dumper Version 14.31.31107.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file crypto-49.lib

File Type: LIBRARY

Archive member name at 8: /               
62C817D6 time/date Fri Jul  8 04:41:10 2022
         uid
         gid
       0 mode
   46DEE size
correct header end

    8851 public symbols

    96AA0 timingsafe_memcmp
    96F44 timingsafe_bcmp
    9725E getentropy
    97700 arc4random_uniform
    97B90 arc4random
    97B90 arc4random_buf
    9A42C explicit_bzero
    9A85C __gmtime_r
```

I might be using the tool wrong, which is why I checked from the source as seen above.

**Comment by Squall-Leonhart** [comment]
use Cmake GUI to create the required solution from the cmake files, using Visual STudio's own cmake implementation will create faulty vc project filters and settings
